site/docs/plugins/i18n.md

```diff
@@ -1,105 +1,751 @@
+
+These two features are going to help you when translating things.
+Fluent provides both.
+You can read their documentation and see how they are used in the following chapter.
+
+[Placeables](https://projectfluent.org/fluent/guide/placeables.html) and [Functions](https://projectfluent.org/fluent/guide/functions.html) are two core features of the Fluent localization system.
+This plugin injects some placeables and functions into your translation context and reduces your work of managing the translation context.
+
+### Placeables
+
+Placeables are used to inject data into a translation string.
+For example, in a welcome message, you may use the user's name to greet them.
+But you can't have messages for each user's name in your translations.
+Instead, you can pass their name in the translation context, and the message will change according to the receiver.
+
+Here is what you would do in `en.ftl`:
+
+```fluent:no-line-numbers
+welcome = Hi there, { $name }!
+```
+
+We can pass a value for the placeable `name`, in the translation context like this:
+
+```ts:no-line-numbers
+await ctx.reply(ctx.t("welcome", { name: ctx.from.first_name }))
+```
+
+Then the message will be different for each user, with their first names.
+
+#### `first_name`
+
+It is a replacement for `ctx.from.first_name`.
+So, you don't need to pass it in the context you need. You can use it directly like this
+
+```fluent:no-line-numbers
+greeting = Hi { $first_name }
+```
+
+### Functions
+
+Functions are like placeables. But unlike placeables, functions get executed and replaced by the value it returns in the final output.
+
+#### `CTX`
+
+This function can help you to access any properties of the bot context directly from your translations.
+
+```fluent:no-line-numbers
+id = Your user ID: { CTX("from.id") }
+```
+
+The `CTX("from.id")` will get replaced by `ctx.from.id`.
+
+Optionally, you can pass in a default value as the second argument.
+If the value is undefined, the default value is used instead.
+
+```fluent:no-line-numbers
+id = Last name: { CTX("from.last_name", "You don't have a last name.") }
+```
+
+## Listening for Localized Text
+
+We managed to send localized messages to the user.
+Now let's take a look at how to listen for messages sent by the user.
+In grammY, we usually use the `bot.hears` handler for listening to incoming messages.
+But since we've been talking about internationalization, in this section we will see how to listen for localized incoming messages.
+
+This feature come in handy when your bot has [custom keyboards](./keyboard.md#custom-keyboards) containing localized text.
+
+Here is a short example of listening to a localized text message sent using a custom keyboard.
+Instead of using `bot.hears` handler, we use `bot.filter` combined with the `hears` middleware provided by this plugin.
+
+```ts
+import { hears } from "@grammyjs/i18n";
+
+bot.filter(hears("back-to-menu-btn"), async (ctx) => {
+  await ctx.reply("main-menu-msg");
+});
+```
```

Comment:
Please provide TS/JS/Deno versions where the import is adjusted so that copying becomes easier.